### ISSUE TITLE
How to prevent Lerna from automatically updating dependencies versions?

### ISSUE THREAD

**Issue by rfgamaral** [lerna/lerna#2326]
For example, say that I have 2 packages configured like this:

**`/packages/foo/package.json`**

```json
{
  "name": "foo",
  "version": "1.0.0"
}
```

**`/packages/bar/package.json`**

```json
{
  "name": "bar",
  "version": "1.0.0",
  "dependencies": {
    "foo": "1.0.0"
  }
}
```

I then make a breaking change to `foo` and on my next `lerna publish` three things will happen:

- `foo` version will be bumped to `v2.0.0` 
- `foo` dependency in `bar` will also be bumped to `v2.0.0`
- `bar` version will be bumped to `v1.0.1` (due to previous point)

**Is there any way to prevent the automatic bump of the `foo` dependency in `bar`?** And, consequentially, prevent the `bar` version bump to `v1.0.1` (nothing changed after all).

**Comment by evocateur** [comment]
If they are sibling packages managed by Lerna, the behavior you describe is intended. (cf #707, etc etc)

If you're trying to force `bar` to stay pinned to `^1.0.0` of `foo`, you can move it back manually and Lerna will stop considering them local siblings (`bar` shouldn't update its dep on `foo` when `foo` changes). At that point, it's questionable why they were monorepo siblings in the first place, but hey, I'm not your real dad.

**Comment by rfgamaral** [comment]
> If you're trying to force `bar` to stay pinned to `^1.0.0` of `foo`, you can move it back manually and Lerna will stop considering them local siblings (`bar` shouldn't update its dep on `foo` when `foo` changes).

But `lerna publish` already tagged and published new versions for both `foo` and `bar` and that's the behavior we're trying to prevent. In this particular scenario we simply want to publish `foo` `v2.0.0` and deal with with the integration of `foo` in `bar` at a later time.

> At that point, it's questionable why they were monorepo siblings in the first place

They are in a monorepo because they belong to the same project and we use `independent` versions mode.

For us it makes sense that there are situations (like a breaking change described above) where we might need to urgently fix something in `foo` but don't have time (nor need) to quickly integrate those breaking changes in `bar`.

**Comment by cspotcode** [comment]
I'm in the same boat as @rfgamaral, trying to puzzle out how best to handle this.  I want CI to publish as automatically as possible so engineers can get canary or pre-release builds to test out their new stuff without the formality of cutting a new release.

It would be nice to set an array in bar's package.json like `"lockedDependencies": ["foo"]`. which tells lerna not to bump the foo dependency in bar.

Even if we *want* bar to be automatically bumped with the new version of foo, which kind of bump is it?  Major, minor, or patch?  A human needs to decide this manually.  A breaking change in foo might not mean a breaking change in bar.

For this reason, maybe it makes the most sense to use a 2-step publishing process with human intervention between the steps.

* Use `--preid` to update the version numbers of foo and bar and update bar's dependency on foo.
* Human intervention: Manually move bar's dependency back to the old version of foo
* Promote and publish the pre-release versions to stable.

This 2-step publishing process allows human intervention to either:

a) keep bar using the old version of foo, or  
b) adjust the bump of bar to be minor or major instead of patch.

CI can do pre-releases automatically, and *only* publish non-pre-release builds when a human manually graduates them.

**Comment by evocateur** [comment]
> They are in a monorepo because they belong to the same project and we use `independent` versions mode.

Independent versioning has absolutely no bearing on the relationship of local packages, it merely allows you to choose different version bumps when changes are detected.

> Even if we want bar to be automatically bumped with the new version of foo, which kind of bump is it? Major, minor, or patch? A human needs to decide this manually. A breaking change in foo might not mean a breaking change in bar.

It defaults to a patch (independently versioned), without any other changes in the dependent that might otherwise influence its own version bump. If the dependent required _no_ changes, then that's a pretty clear signal (imo) that a major bump for the dependent is not necessary. If that's not the case, then perhaps the test suite needs some beefing up? One can argue that the dependent whose dependency major changes might justify a minor bump instead of a patch, but that's semantics (lol).

One of the benefits of using a monorepo is the ability to coordinate API changes across many packages at the same time. Many, some, or none of the changes might necessitate a major version bump, but aside from mind-reading, Lerna really doesn't have anything else to go off of (other than _what_ changed). Attempting to automate such complicated logic in CI is really trying to fit a square peg in a round hole with Lerna as it is today. It was built for manual execution from a dev's localhost, and most of the design decisions reflect that.

**Comment by hendrikvolschenk** [comment]
`package.json` already indicates what it allows to install up to:

* `>1.0.0` - Bump everything, including major.
* `~1.0.0` - Bump only patch versions.
* `^1.0.0` - Bump patch and minor versions.
* `1.0.0` - Don't touch anything

If `lerna` could respect these when bumping other packages (while keeping/transferring the prefix)
then it would work for many of us with the same needs as @rfgamaral 
and allow for even more control in other situations.

**Comment by Jasperrr91** [comment]
Yes @hendrikvolschenk. Lerna should stick to the guidelines provided by `package.json`.

Today, we explicitly pinned a dependency to the last version to not update it while pushing through a breaking change on another package, yet Lerna still bumped the dependency. This is a very frustrating and honestly, pretty bad, 'functionality'.

**Comment by BrianLoughnane** [comment]
+1 on this issue. Would love if sibling dep version ranges were respected (ie overwritten by a new range only if the new version matches the existing range). Seems that the current behavior is redundant to the "file:path/to/package" specifier but allows no such specification of sibling dep version ranges.

**Comment by cpertile** [comment]
+1 on this. An option to bump (or not) dependents would be great for the team I work in. Sometimes we have to release a new dependency version but every package that uses that dependency also gets bumped just because of lerna updating the `package.json` files troughout the other packages. In our workflow this just creates a bunch of commits/tags with a change in `package.json` that we don't want/need.

When we build our apps, it will already install the latest version of the dependency (it's listed as `dependency: ^1.0.2`. So there's no point in bumping versions for us.

**Comment by Dante-dan** [comment]
+1 on this issue. `lerna` should respect these when bumping other packages

**Comment by fjodor-rybakov** [comment]
+1

**Comment by Dante-dan** [comment]
I want add this feature, but lerna is difficult to debug. 😭

**Comment by arsen-melkumiants** [comment]
+1

**Comment by TheHunterDog** [comment]
+1

**Comment by AxelPeter** [comment]
+1

**Comment by ms2d** [comment]
This does not make any sense to me. If `packageA` depends on `packageB@1.0.0`, and I later bump B to `v2.0.0`, nothing has changed for `packageA` (it still depends on `packageB@1.0.0`). So why would `packageA's` version number get bumped as well?

**Comment by aka-toxa** [comment]
+1

**Comment by khaelys** [comment]
+1

**Comment by mitorn** [comment]
+1

**Comment by chkalovec-spec** [comment]
+1, any updates for this?

**Comment by sivaraj-v** [comment]
Hope this script will solve the issue, untill the lerna gives the feature.
Update : `const packagesDir = path.join(__dirname, 'packages');` as you your project

https://gist.github.com/sivaraj-v/f6556b2412a7073fc76146fefd85dc26

**Comment by iduspara-ftip** [comment]
Did anyone find a solution for this, except having a custom script which will update the version?
